### DISCUSSION TITLE
Could async export be allowed/enabled from config.js?

### DISCUSSION THREAD

**Discussion by pmorch** [renovatebot/renovate#13033] [Suggest an Idea]
In order to work around #10011 ("Allow autodiscover filtering for repo topic "), I was hoping to use [`got`](https://www.npmjs.com/package/got) to get a list of the repositories for a particular topic, and provide that in the `repositories` value for the exported config. Because `got` and (all indeed HTTP libs) operate asynchronously, `config.js` cannot export its config synchronously if that is to work.

And when `module.exports` exports a `Promise`, the config is not recognized.

Would a patch like this be considered (with corresponding tests to `file.spec.ts` that test both vanilla config and promises)?

```diff
diff --git a/lib/workers/global/config/parse/file.ts b/lib/workers/global/config/parse/file.ts
index 103d24d86..b209c473b 100644
--- a/lib/workers/global/config/parse/file.ts
+++ b/lib/workers/global/config/parse/file.ts
@@ -18,7 +18,10 @@ export async function getParsedContent(file: string): Promise<RenovateConfig> {
       return JSON5.parse(await readFile(file, 'utf8'));
     case '.js': {
       const tmpConfig = await import(file);
-      return tmpConfig.default ? tmpConfig.default : tmpConfig;
+      let config = tmpConfig.default ? tmpConfig.default : tmpConfig;
+      // Allow the config to be a promise
+      config = await config;
+      return config;
     }
     default:
       throw new Error('Unsupported file type');
```
(patch edited since original post due to conflicts)

Because then  a `config.js` like this works to implement #10011 :

```javascript
const got = require('got')

const endpoint = 'https://gitlab.example.com/api/v4/'
const token = 'mytoken'

// Get repositories matching this topic:
const topic = "helm"

async function getRepositories() {
    // Need to handle paging when more than 100 repos...
    const result = await got(endpoint + '/projects', {
        searchParams: {
            'topic': topic,
            'per_page': 100,
        },
        headers: {
            'PRIVATE-TOKEN': token
        },
        responseType: 'json',
    });

    const repositories = result.body.map(r => { return r.path_with_namespace });
    return repositories
}

module.exports = (async () => {
    return {
        token,
        endpoint,
        platform: 'gitlab',
        onboarding: false,
        dryRun: true,
        repositories: await getRepositories()
    }
})()
```

**Comment by viceice** [comment]
I like the idea to allow to export a config function which can be return a promise

**Reply by pmorch** [reply]
@viceice : Can I interpret "you like the idea" as: "Go ahead and make a pull request"?

Just to be clear: My suggestion is not to export a config function which can return a `Promise`, but to export the `Promise` directly. The anonymous `async` function in my example above is called (with a trailing `()`).

**Reply by viceice** [reply]
see https://github.com/renovatebot/renovate/issues/13035

Please wait for confirm of other maintainers.

**Comment by pmorch** [comment]
At 13:45:48 2021 +0100 today, two hours before I posted this, @nejch added 9aa97af5b346f82b48564c20c66783b51c3022d9 to main that created a merge conflict with my original patch above :-). So I updated it so it applies to current main.

I just didn't want anybody to wonder why my patch didn't apply....
